readme.txt

Desquirr - A decompiler plugin for Interactive Disassembler Pro
===============================================================

Copyright (c) 2002 David Eriksson <david@2good.nu>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


Introduction
------------

Desquirr is a decompiler plugin for IDA Pro. It is currently capable of simple
data flow analysis of binaries with Intel x86 machine code.

This program is currently under development. Suggestions, bug reports and
patches are welcome.

For more information, please visit this web site:

  http://desquirr.sourceforge.net/

  
Installation and configuration
------------------------------

1. Copy desquirr.plw to the plugins subdirectory in your IDA Pro installation

2. Edit plugins.cfg (optionally) to get different kinds of decompilation:

     Decompile_to_C desquirr Ctrl+F9  0
     Decompile desquirr Ctrl+F10  1
              

Limitations
-----------

The limitations are not limited to this list :-)

o Only decompiles one function at a time

o Does not handle all types of instructions

o Does not handle DX:AX or EDX:EAX very well

o Does not handle register variables properly

o Leaves pop statements in the end of functions

o Lots of code is inlined in the .hpp files but should be moved to the .cpp
  files


Software needed to compile Desquirr
-----------------------------------

Interactive Disassembler Pro and SDK version 4.21:

  http://www.datarescue.com/idabase/

The Boost C++ Libraries version 1.27.0:

  http://boost.org/

The free Borland C++Builder Compiler:

  http://www.borland.com/bcppbuilder/freecompiler/




docs/ChangeLog
Tue Jan 30 11:42:30 WEST 2007

- fixed truncated ida message display output
- improved expression readability, by writing brackets only when needed according to the operator precendence rules
- arm decompiler now assumes r0 .. r3 are the parameters.
- improved handling of arm shift operators
- arm: now ignoring constant pools embedded in code
- added support for several missing arm opcodes: ldrpc, addsp, condition codes
- fixed some nullptr ref bugs
- added msvc8 project files
- added gnumake makefile to build with msvc8
- improved global variable / label naming

Mon Oct 17 13:02:05 WEDT 2005

- fixed arm 'S' handling: changed && to &

Sun Oct 16 08:17:34 WEDT 2005

- made ida 4.9 compatible
- fixed  bug where push #imm was counted with x86 func prolog.
- added support for arm:MOVL macro instruction
- improved EscapeAsciiString
- added flag to process all functions in a file.
- made to compile ( but not yet run ) with gcc:
     * added virtual destructors
     * split declaration / init of PLUGIN.
     * added some ifndef GCC around watcom specific stuff.
- renamed 'Call' to 'CallExpression'
- renamed 'Global' to 'GlobalVariable' 
- added dump functions for nodes, instructions, expressions.
- changed several const char* to std::string
- added some warnings of unexpected conditions
- GlobalVariable::GetName now generates name when name not found.
- added several 'const' modifiers.
- changed several std::string (byval) params, into const std::string&. (byref)

Fri Jul 23 WEDT 2005

- split off x86 stuff into 'ida-x86.cpp'
- renamed LowLevelAnalysis to X86Analysis
- changed deprecated #include <strstream> to #include <sstream>
- removed 'using namespace std;' directives, .. added 'std::' where needed.
- added idaapi to the public functions, for ida4.80 compatibility
- type inconsistancy in 'CreateFrom', GetString, GetName in expression.hpp /
expression.cpp, changed Addr to ea_t
- moved *StackVar* function decls to idapro.hpp
- removed use of strcpy from RegisterName, and ConditionOp
- use same register numbering scheme as arm processor module
- use named constants for register index.
- fixed crash, when isStkVar(flags, 1) and isStkVar(flags, 2) are both true.
  while only operand==2 is actually the right one.
- changed snprintf into qsnprintf
- changed GetStackVariableName to early-exit style
- more return code checking
- fixed call to get_ti to be ida4.80 compatible
- added 'const' and &reference to several BoolArray methods, so it compiles with gcc too.
- changed constructor initializer order to reflect real init order.
  in Location()
- changed 'const char* const RegisterName()' to 'std::string RegisterName() const'

Fri Jul 22 21:20:28 WEDT 2005

- no compiles both for ida470 and ida480
- MLA now decompiles correctly
- added several warnings for incorrectly handled cases

TODO:
- register index is now a mix of int, unsigned short, short.
- INVALID_ADDR should be of type Addr
- include stl headers before ida's. ... otherwise strcpy will be redefined and
  compile fails under gcc.

- a situation like this is not handled properly:
   l1:
     push #1
	 jmp L3
   l2:
     push #2
   l3:
     call fn

- change (reg<<16)>>16  to typecast: (WORD)reg
- add code to simplify expressions.

- def/use chains should also view local vars + globals.

- TODO: i think i should make a CallExpression object, instead of an
   assignment(reg, call(params)) and handle exprs+return regs. later
- TODO: remark all default params from the .hpp in the .cpp files.
- TODO: review all "" and NULL ... change to something sensible ... like throw
   exception
- TODO; change most 'return false'  to throw exception.
- TODO: 'xyz_type' and 'switch'  should not be used. ... everything should be
	solved virtually.
- TODO: check if it is better to return const& than return value?

- TODO:
    * make Register, Location, StringLiteral subclass of Variable
    * change char register to type Location
	* change BoolArray, reg2addr_map  to VariableSet, Variable2addr_map
	   Instruction::Uses(Variable var)  returns true if ins uses var.
	      ( etc. also for Defines, etc.  )
- change static_cast's  to dynamic_cast's

- shared_from_this
   * only on the leaf nodes of the class hierarchy.
   * not in destructor/constructor
   * base classes need to be abstract

- get rid of Frontend::Get construction
- what is the 'erasepool' good for?

- need some kind of way to note that a change to AL, also changes AX, and EAX.
   but not AH.

- Instruction::Return  should not fix the return register.
   ... it should be derived from du-analysis of entir func.

- investigate o_mem decoding in IdaArm
- arm: add Rx, BP, #imm  should be reference to stackver.
- arm: sub Rx, BP, #imm  should be reference to stackver.
- seemingly unused procedure params now get deleted, 
- seemingly unused array members now get deleted.
- in several cases the condition code is not checked.
- add support for structs.

FIXED:
- arm: add Rx, SP, #imm   now translates to reference to stackvar.
- array index is now calculated properly for stackvars.


idapro.cpp
// 
// Copyright (c) 2002 David Eriksson <david@2good.nu>
// 
// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the "Software"),
// to deal in the Software without restriction, including without limitation
// the rights to use, copy, modify, merge, publish, distribute, sublicense,
// and/or sell copies of the Software, and to permit persons to whom the
// Software is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
// DEALINGS IN THE SOFTWARE.
//
// $Id: idapro.cpp,v 1.12 2007/01/30 09:49:12 wjhengeveld Exp $
#include "idainternal.hpp"
#include "idapro.hpp"
#include "instruction.hpp"
#include "analysis.hpp"

#include <memory>

#if IDP_INTERFACE_VERSION<76
// backward compatibility with ida480
size_t get_member_name(tid_t mid, char *buf, size_t bufsize)
{
    char *name= get_member_name(mid);
    if (name) {
        qstrncpy(buf, name, bufsize);
        return strlen(buf);
    }
    return -1;
}
#endif

#if IDP_INTERFACE_VERSION<75
// backward compatibility with ida470
bool get_ti(ea_t ea, type_t *buf, size_t bufsize, p_list *fnames, size_t fnsize)
{
    return get_ti(ea, buf, fnames);
}
#endif
std::string get_struct_path(struc_t *struc, int offset, int *pIndex)
{
	member_t* member = NULL;  /* must live here, not inside while */
	std::ostringstream buffer;
	bool first = true;

	// todo: this is not working as intended. needs fixing.
	int totalsoff= 0;
	while (struc)
	{
		member = get_member(struc, offset);
		if (member)
		{
			if (first)
				first = false;
			else
				buffer << '.';

			char tmp[256];
			if (get_member_name(member->id, tmp, 256)>0)
				buffer << tmp;
			else
				buffer << "NO_NAME";

			totalsoff = offset - member->soff;
			struc = get_sptr(member);
		}
		else
		{
			struc = NULL;
		}
	}

	//if (member && (offset - member->soff) != 0)
	if (totalsoff)
	{
		//message("%p Stack variable: offset=%i, member->soff=%i\n", ea, offset, member->soff);
		if (pIndex)
			*pIndex = totalsoff; // offset - member->soff;
	}

	return buffer.str();
}/*}}}*/

// used from CreateStackVariable
std::string GetStackVariableName(const insn_t &insn, int operand, int* pIndex)/*{{{*/
{
	if (pIndex)
		*pIndex = 0;
	
    ea_t ea= insn.ea;
	func_t* func = get_func(ea);
	if (func==NULL)
	{
		message("ERROR - get_func(%08lx)\n", ea);
		return "";
	}


	ulong offset = calc_stkvar_struc_offset(func, ea, operand);
	if (offset==BADADDR) {
		message("ERROR in calc_stkvar_struc_offset(%08lx, %08lx, %d)\n", func->startEA, ea, operand);
		return "";
	}

	struc_t* struc = get_frame(func);
	if (struc==NULL) {
		message("ERROR: function has no frame\n");
		return "";
	}
	return get_struct_path(struc, offset, pIndex);
}


// returns name such that get_func_name(ea) == name and get_func(ea).startEA+*pIndex == ea
// used by GetGlobalVariableName
std::string GetGlobalCodeLabel(ea_t ea, int *pIndex)/*{{{*/
{
    // return funcname + offset
    char name[MAXSTR];

    func_t *func= get_func(ea);
    if (func==NULL) {
        message("%p Warning: referenced code offset not in a function\n", ea);
        return "";
    }
    if (get_func_name(ea, name, MAXSTR)) {
        *pIndex= ea - func->startEA;
        return std::string(name);
    }
    else {
        message("%p Warning: referenced code offset not in a function\n", ea);
        return "";
    }
}

// returns name such that get_name_ea(ea - *pIndex) == name
std::string GetLocalCodeLabel(ea_t ea, int *pIndex)/*{{{*/
{
    char name[MAXSTR];

    func_t* pfn= get_func(ea);
    func_item_iterator_t fii;
    for ( bool ok=fii.set(pfn, ea); ok; ok=fii.prev_addr() ) {
        ea_t lea = fii.current();

        if (get_name(ea, lea, name, sizeof(name))) {
            *pIndex= ea-lea;
			if (*pIndex)
				message("Unexpected locallabel with name=%s index=%d\n", name, *pIndex);
            return std::string(name);
        }
		else {
			break;
			// message("%p %p -- fii next\n", ea, lea);
		}
    }
      
    message("%p Warning: label without name\n", ea);
    return "";
}/*}}}*/

Instruction_ptr CreateLocalCodeLabel(ea_t ea)
{
    int index;
    std::string name= GetLocalCodeLabel(ea, &index);
    if (index || name.empty()) {
        name.resize(32);
        name.resize(qsnprintf(&name[0], name.size(), "loc_%X", ea));
        message("NOTE: created new label %s\n", name.c_str());
    }
    Instruction_ptr instr;
    // todo: think of a better way to represent local function labels.
    instr.reset(new Label(ea, name.c_str()));
    return instr;
}

Expression_ptr CreateLocalCodeReference(ea_t ea)
{
    int index;
    std::string name= GetLocalCodeLabel(ea, &index);
    if (index || name.empty()) {
        name.resize(32);
        name.resize(qsnprintf(&name[0], name.size(), "loc_%X", ea));
        message("NOTE: created new label %s\n", name.c_str());
    }
    Expression_ptr expr;
    // todo: think of a better way to represent local function labels.
    expr.reset(new GlobalVariable(name, 0, ea));
    return expr;
}
Expression_ptr CreateGlobalCodeLabel(ea_t ea)
{
    int index;
    std::string name= GetGlobalCodeLabel(ea, &index);
    if (name.empty()) {
        name.resize(32);
        name.resize(qsnprintf(&name[0], name.size(), "proc_%X", ea));
        message("NOTE: created new function name %s\n", name.c_str());
    }
    else if (index!=0) {
        name.resize(name.size()+16);
        name.resize(qsnprintf(&name[0], name.size(), "%s+0x%X", name.c_str(), index));
        message("NOTE: using func+offs name: %s\n", name.c_str());
    }
    Expression_ptr expr;
    expr.reset(new GlobalVariable(name, 0, ea));
    return expr;
}

// todo: figure out how to get the structure type of the data at a specific offset.
std::string GetGlobalVariableName(ea_t ea, int* pIndex)/*{{{*/
{
	std::ostringstream buffer;

	if (pIndex)
		*pIndex = 0;

    flags_t flags= ::getFlags(ea);

    if (isCode(flags)) {
        return GetGlobalCodeLabel(ea, pIndex);
    }
/*
    else if (!isData(flags)) {
        message("%p Warning: referencing unknown item flags=%08lx\n", ea, flags);
    }
    else if (isStruct(flags)) {
        tid_t tid= 
        struc_t* struc = get_struc(tid);
        // get_stroff_path(ea_t ea, int n, tid_t *path, adiff_t *delta)
    }
*/
    else {
        char name[MAXSTR];
        if (get_name(BADADDR, ea, name, sizeof(name))) {
            return name;
        }
        else {
			ea_t head= prev_head(ea, 0);

			std::string headname; headname.resize(MAXSTR);
			if (!get_name(BADADDR, head, &headname[0], headname.size())) {
				headname.resize(qsnprintf(&headname[0], headname.size(), "gvar_%X", head));
			}
			else {
				headname.resize(strlen(&headname[0]));
			}
			tid_t tid= get_strid(head);
			struc_t *struc= get_struc(tid);
			return headname + "." + get_struct_path(struc, ea-head, pIndex);
        }
    }
}/*}}}*/
bool is_local_to_function(ea_t funcea, ea_t ea)
{
	func_t *func= get_func(funcea);
	return func_contains(func, ea);
}
Expression_ptr CreateVariable(const insn_t &insn, int operand)
{
	ea_t ea= insn.Operands[operand].addr;

	if (is_local_to_function(insn.ea, ea))
		return CreateLocalCodeReference(ea);
	else
		return CreateGlobalVariable(insn, operand);
}
Expression_ptr CreateGlobalVariable(const insn_t &insn, int operand)
{
    Expression_ptr expr;

    ea_t ea= insn.Operands[operand].addr;

    int index;
    std::string name= GetGlobalVariableName(ea, &index);

    if (name.empty()) {
        name.resize(32);
        name.resize(qsnprintf(&name[0], name.size(), "gvar_%X", ea));
        message("NOTE: created new globalvar %s\n", name.c_str());
    }

    expr.reset(new GlobalVariable(name, index, ea));

    return expr;
}



Expression_ptr CreateStackVariable(insn_t& insn, int operand)/*{{{*/
{
	Expression_ptr result;
	
	int index;
    std::string name = GetStackVariableName(insn, operand, &index);

	if (name.empty())
	{
		// Try to add a stack variable and try again!
//		message("%p Warning: trying to create stack variable\n", insn.ea);
		if (!add_stkvar(insn.Operands[operand], insn.Operands[operand].addr)) {
			message("error in add_stkvar(%08lx, %08lx)\n", insn.Operands[operand].dtyp, insn.Operands[operand].addr);
			return Expression_ptr();
		}
		if (!op_stkvar(insn.ea, operand)) {
			message("error in op_stkvar(%08lx, %08lx)\n", insn.ea, operand);
			return Expression_ptr();
		}
		name = GetStackVariableName(insn, operand, &index);
	}
	
	if (!name.empty()) {
		result.reset(new StackVariable(name, index));
	}
	else {
		message("ERROR: could not allocate stack var (%08lx, %d)\n", insn.ea, operand);
	}

	return result;
}/*}}}*/


static const char* const optype_string[] = {/*{{{*/
	"o_void",
	"o_reg",
	"o_mem",
	"o_phrase",
	"o_displ",
	"o_imm",
	"o_far",
	"o_near",
	"o_idpspec0",
	"o_idpspec1",
	"o_idpspec2",
	"o_idpspec3",
	"o_idpspec4",
	"o_idpspec5",
	"o_last"
};/*}}}*/

const char* IdaPro::GetOptypeString(op_t& op)/*{{{*/
{
//	if (op.type >= 0 && op.type <= o_last)
		return optype_string[op.type];
/*	else
		return "INVALID";*/
}/*}}}*/

insn_t GetLowLevelInstruction(ea_t address)/*{{{*/
{
	// note: ua_ana0  sets the global 'cmd' variable
	ua_ana0(address);
	return cmd;
}/*}}}*/

void IdaPro::DumpInsn(Addr address)/*{{{*/
{
	insn_t insn = ::GetLowLevelInstruction(address);
	DumpInsn(insn);
}/*}}}*/


int IdaPro::vmsg(const char *format, va_list va)
{
	return ::vmsg(format, va);
}

Addr IdaPro::AddressFromName(const char *name, Addr referer)
{
	return ::get_name_ea(referer, name);
}

void IdaPro::LoadCallTypeInformation(CallExpression* call)
{
	if (BADADDR == call->Address())
		return;

	type_t type[MAXSTR];
	p_list names[MAXSTR];
	
	if (!get_ti(call->Address(), type, MAXSTR, names, MAXSTR))
  {
    message("No type information for function at %p!\n", 
        call->Address());
		return;
  }

	// CM (calling convention & model)
	cm_t cm = type[1];
	call->CallingConvention(cm);

	ulong plocations[CallExpression::MAX_PARAMETERS];
	memset(plocations, 0, sizeof(plocations));

	type_t* DataTypes     [CallExpression::MAX_PARAMETERS];
	char*   ParameterNames[CallExpression::MAX_PARAMETERS];
	memset(DataTypes,      0, sizeof(DataTypes));
	memset(ParameterNames, 0, sizeof(ParameterNames));

	call->ParameterCount(
			build_funcarg_arrays(type, names, plocations, 
			(type_t**)DataTypes, (char**)ParameterNames, CallExpression::MAX_PARAMETERS, false));

#if 0
	char buffer[MAXSTR]; 
	for (int i = 0; i < mParameterCount; i++)
	{
		buffer[0] = '\0';
		print_type_to_one_line(buffer, sizeof(buffer), idati, mDataTypes[i]);
		msg("Parameter %i type: %s %s\n", i, buffer, mParameterNames[i]);
	}
#endif

	type_t return_type[MAXSTR];
	if (::extract_func_ret_type(type, return_type, sizeof(return_type)))
	{
		//DataType().Set(return_type);
#if 0
		buffer[0] = '\0';
		print_type_to_one_line(buffer, sizeof(buffer), idati, mReturnType);
		msg("Return type: %s\n", buffer);
#endif
	}

	free_funcarg_arrays(DataTypes, ParameterNames, CallExpression::MAX_PARAMETERS);
	
}

